airflow/utils/metastore_cleanup.py

```diff
@@ -0,0 +1,321 @@
+
+    if keep_last_group_by is not None:
+        subquery = subquery.group_by(keep_last_group_by)
+
+    subquery = subquery.from_self()
+    return subquery
+
+
+def _build_query(
+    *,
+    orm_model,
+    recency_column,
+    keep_last,
+    keep_last_filters,
+    keep_last_group_by,
+    clean_before_timestamp,
+    session,
+    **kwargs,
+):
+    query = session.query(orm_model)
+    conditions = [recency_column < clean_before_timestamp]
+    if keep_last:
+        subquery = _subquery_keep_last(
+            recency_column=recency_column,
+            keep_last_filters=keep_last_filters,
+            keep_last_group_by=keep_last_group_by,
+            session=session,
+        )
+        print(subquery.all())
+        conditions.append(recency_column.notin_(subquery))
+    query = query.filter(and_(*conditions))
+    return query
+
+
+logger = logging.getLogger(__file__)
+
+
+def _cleanup_table(
+    *,
+    orm_model,
+    recency_column,
+    keep_last,
+    keep_last_filters,
+    keep_last_group_by,
+    clean_before_timestamp,
+    dry_run=True,
+    verbose=False,
+    session=None,
+    **kwargs,
+):
+    print()
+    if dry_run:
+        print(f"Performing dry run for table {orm_model.__tablename__!r}")
+    query = _build_query(
+        orm_model=orm_model,
+        recency_column=recency_column,
+        keep_last=keep_last,
+        keep_last_filters=keep_last_filters,
+        keep_last_group_by=keep_last_group_by,
+        clean_before_timestamp=clean_before_timestamp,
+        session=session,
+    )
+
+    _print_entities(query=query, print_rows=False)
+
+    if not dry_run:
+        _do_delete(query=query, session=session)
+        session.commit()
+
+
+def _confirm_delete(*, date: DateTime, tables: List[str]):
+    for_tables = f" for tables {tables!r}" if tables else ''
+    question = '\n'.join(
+        [
+            f"You have requested that we purge all data prior to {date}{for_tables}.",
+            "This is irreversible.  Consider backing up the tables first and / or doing a dry run"
+            " with option --dry-run.",
+            "Enter 'delete rows' to proceed.",
+        ]
+    )
+    print(question)
+    answer = input().strip()
+    if not answer == 'delete rows':
+        raise SystemExit("User did not confirm; exiting.")
+
+
+def _print_config(*, configs: Dict[str, _TableConfig]):
+    data = [x.readable_config for x in configs.values()]
+    AirflowConsole().print_as_table(data=data)
+
+
+class _warn_if_missing(AbstractContextManager):
+    def __init__(self, table, suppress):
+        self.table = table
+        self.suppress = suppress
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exctype, excinst, exctb):
+        caught_error = exctype is not None and issubclass(exctype, OperationalError)
+        if caught_error:
+            logger = logging.getLogger()
+            logger.warning(f"Table {self.table!r} not found.  Skipping.")
```

Comment:
```suggestion
            logger.warning(f"Table %r not found.  Skipping.", self.table)
```

Don’t use f-string in logging in general.